app/server/ruby/bin/daemon.rb

```diff
@@ -0,0 +1,832 @@
+#                       the Spider Server.
+#
+# scsynth:              UDP port on which scsynth listens (necessary for
+#                       connecting to the boost shared memory for scope data)
+#
+# osc-cues:             UDP port used to receive OSC cue messages from external
+#                       processes.
+
+module SonicPi
+  module Daemon
+
+    class Init
+
+      def initialize
+
+        @safe_exit = SafeExit.new do
+          # Register exit routine
+          # This will only be called once
+          Util.log "Daemon Booter is now exiting."
+          Util.log "Cleaning up any running processes..."
+          cleanup_any_running_processes
+          Util.log "Daemon Booter - Over and Out."
+          Util.close_log
+        end
+
+        # This is where the Daemon begins and ends.
+        @scsynth_booter  = nil
+        @tau_booter      = nil
+        @spider_booter   = nil
+
+        clear_logs
+
+        Util.open_log
+        Util.log "Welcome to the Daemon Booter"
+
+        ports = PortDiscovery.new(@safe_exit).ports
+
+        Util.log "Selected ports: "
+        Util.log ports.inspect
+
+        @zombie_kill_switch = spawn_zombie_kill_switch(ports["daemon-keep-alive"])
+
+        # Let the calling process (likely the GUI) know which port to
+        # listen to and communicate on with the Ruby spider server via
+        # STDOUT.
+        puts "#{ports["daemon-keep-alive"]} #{ports["gui-listen-to-server"]} #{ports["gui-send-to-server"]} #{ports["scsynth"]} #{ports["osc-cues"]}"
+        STDOUT.flush
+
+        # Boot processes
+
+        Util.log "Booting Scsynth"
+        @scsynth_booter = ScsynthBooter.new(ports)
+
+        Util.log "Booting Tau"
+        @tau_booter     = TauBooter.new(ports)
+
+        Util.log "Booting Spider Server"
+        @spider_booter  = SpiderBooter.new(ports)
+
+        # Wait for processes
+        @tau_booter.wait if @tau_booter
+        Util.log "Tau process has completed"
+
+        @scsynth_booter.wait if @scsynth_booter
+        Util.log "Scsynth process has completed"
+
+        @spider_booter.wait if @spider_booter
+        Util.log "Spider Server process has completed"
+      end
+
+
+      # This is the Zombie Kill Switch
+      #
+      # It is a separate thread which calls exit (therefore triggering
+      # the at_exit block which in turn calls
+      # cleanup_any_running_processes).
+      #
+      # There are the following exit conditions:
+      #
+      # Initial Connection Timeout - if nothing connects within 5 seconds,
+      #                              switch is activated
+      #
+      # Connection Lost            - if the connection is broken or there's
+      #                              a timeout - waiting for the next
+      #                              keep-alive message.
+      #
+      def spawn_zombie_kill_switch(port_num, &blk)
+        # Define how long to wait for an initial connection and between keep-alive messages.
+        connect_timeout = 5
```

Comment:
Further above you mention in comments that the keep alive time out is 3 seconds. That might need updating?